src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/service/TimelineEventFactory.kt

```diff
@@ -0,0 +1,131 @@
+
+  fun goalCreatedTimelineEvent(goal: Goal) =
+    buildTimelineEvent(goal, goal.reference, TimelineEventType.GOAL_CREATED)
+
+  /**
+   * Determines what type of [TimelineEvent]s have occurred, following an update to a [Goal]. For example, whether its title
+   * has changed, or one of its [Step]s has been completed.
+   *
+   * @param previousGoal The state of the [Goal] just before it was updated.
+   * @param updatedGoal The newly updated [Goal] containing the latest changes.
+   */
+  fun goalUpdatedEvents(previousGoal: Goal, updatedGoal: Goal): List<TimelineEvent> {
+    val events = mutableListOf<TimelineEvent>()
+    // check if the goal itself was updated, excluding its steps and status
+    if (hasGoalBeenUpdated(previousGoal = previousGoal, updatedGoal = updatedGoal)) {
+      events.add(
+        buildTimelineEvent(
+          goal = updatedGoal,
+          sourceReference = updatedGoal.reference,
+          eventType = TimelineEventType.GOAL_UPDATED,
+        ),
+      )
+    }
+
+    if (hasGoalStatusChanged(previousGoal = previousGoal, updatedGoal = updatedGoal)) {
+      val eventType = getGoalStatusEventType(updatedGoal)
+      events.add(
+        buildTimelineEvent(
+          goal = updatedGoal,
+          sourceReference = updatedGoal.reference,
+          eventType = eventType,
+        ),
+      )
+    }
+
+    // check if the status of any steps was changed
+    updatedGoal.steps.forEach {
+      val previousStep = getPreviousStep(previousGoal.steps, it)
+      if (hasStepBeenUpdated(previousStep = previousStep, updatedStep = it)) {
+        events.add(
+          buildTimelineEvent(
+            goal = updatedGoal,
+            sourceReference = previousStep!!.reference,
+            eventType = TimelineEventType.STEP_UPDATED,
+          ),
+        )
+      }
+      if (hasStepStatusChanged(previousStep = previousStep, updatedStep = it)) {
+        val eventType = getStepStatusEventType(it)
+        events.add(
+          buildTimelineEvent(
+            goal = updatedGoal,
+            sourceReference = previousStep!!.reference,
+            eventType = eventType,
+          ),
+        )
+      }
+    }
+    return events
+  }
+
+  private fun hasGoalBeenUpdated(previousGoal: Goal, updatedGoal: Goal) =
+    updatedGoal.title != previousGoal.title ||
+      updatedGoal.steps.size != previousGoal.steps.size ||
+      updatedGoal.lastUpdatedAtPrison != previousGoal.lastUpdatedAtPrison ||
+      updatedGoal.reviewDate != previousGoal.reviewDate ||
+      updatedGoal.notes != previousGoal.notes
+
+  private fun hasGoalStatusChanged(previousGoal: Goal, updatedGoal: Goal) =
+    updatedGoal.status != previousGoal.status
+
+  private fun hasStepBeenUpdated(previousStep: Step?, updatedStep: Step) =
+    previousStep != null && (
+      updatedStep.title != previousStep.title ||
+        updatedStep.targetDateRange != previousStep.targetDateRange ||
+        updatedStep.sequenceNumber != previousStep.sequenceNumber
```

Comment:
I'm in two minds about including `sequenceNumber` here. If say two steps are re-ordered then this will result in two 'STEP_UPDATED' events, which may not seem right.

Having said that, the UI can always decide to filter this and perhaps just report that 'a' step was updated.